Error with BatchNorm conversions in training mode

When using ONNX models with BatchNorm layers saved in training mode, during the keras conversion, an error `ValueError: name for name_scope must be a string` occurs.

Use case: export PyTorch model to onnx with `torch.onnx.export` and argument `training=torch.onnx.TrainingMode.TRAINING`, and then convert the onnx model to keras with `onnx_to_keras`. This is necessary if e.g. we want to fine-tune the target model in keras.

This happens on conversion of normalization layers (observed on BatchNorm only, by should also be observed with InstanceNorm).

Minimal example: converts a ResNet18 model from torch in training mode:
```python
import io

import numpy as np
import onnx
import torch
from torchvision.models import resnet18



if __name__ == "__main__":

    # instanciate PyTorch model
    model = resnet18()
    model.train()

    # create placeholder input tensors
    input_np = np.random.randn(1, 3, 224, 224)
    input_t = torch.from_numpy(input_np).float()

    # export PyTorch model to ONNX
    input_names = ["input_1"]
    output_names = ["output_1"]

    temp_f = io.BytesIO()
    torch.onnx.export(
        model,
        input_t,
        temp_f,
        verbose=True,
        input_names=input_names,
        output_names=output_names,
        opset_version=14,
        do_constant_folding=False,
        dynamic_axes={
            "input_1": {0: "batch_size"},
            "output_1": {0: "batch_size"},
        },
        training=torch.onnx.TrainingMode.TRAINING,
    )
    temp_f.seek(0)
    onnx_model = onnx.load(temp_f)

    # check the ONNX model
    onnx.checker.check_model(onnx_model)

    # export ONNX model to TF Keras
    k_model = onnx_to_keras(
        onnx_model,
        input_names=input_names,
        verbose=True,
        change_ordering=True,
    )
```

Here is a pytest-formatted traceback of the error:

```python
    k_model = onnx_to_keras(
onnx2keras/converter.py:176: in onnx_to_keras
    AVAILABLE_CONVERTERS[node_type](
onnx2keras/normalization_layers.py:56: in convert_batchnorm
    layers[node_name] = bn(input_0)
../../.local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:925: in __call__
    return self._functional_construction_call(inputs, args, kwargs,
../../.local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:1095: in _functional_construction_call
    with graph.as_default(), backend.name_scope(self._name_scope()):
../../.local/lib/python3.8/site-packages/tensorflow/python/keras/backend.py:912: in name_scope
    return ops.name_scope_v2(name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tensorflow.python.framework.ops.name_scope_v2 object at 0x7f971a87adc0>, name = ['122', '123', '124', '/']

    def __init__(self, name):
      """Initialize the context manager.

      Args:
        name: The prefix to use on all names created within the name scope.

      Raises:
        ValueError: If name is not a string.
      """
      if not isinstance(name, six.string_types):
>       raise ValueError("name for name_scope must be a string.")
E       ValueError: name for name_scope must be a string.

../../.local/lib/python3.8/site-packages/tensorflow/python/framework/ops.py:6593: ValueError
```

--- a/onnx2keras/normalization_layers.py
+++ b/onnx2keras/normalization_layers.py
@@ -16,6 +16,8 @@ def convert_batchnorm(node, params, layers, lambda_func, node_name, keras_name):
     :param keras_name: resulting layer name
     :return: None
     """
+    if isinstance(keras_name, list):
+        keras_name = keras_name[0]
     logger = logging.getLogger('onnx2keras.batchnorm2d')
 
     input_0 = ensure_tf_type(layers[node.input[0]], name="%s_const" % keras_name)
@@ -67,6 +69,8 @@ def convert_instancenorm(node, params, layers, lambda_func, node_name, keras_nam
     :param keras_name: resulting layer name
     :return: None
     """
+    if isinstance(keras_name, list):
+        keras_name = keras_name[0]
     logger = logging.getLogger('onnx2keras.instancenorm2d')
 
     input_0 = ensure_tf_type(layers[node.input[0]], name="%s_const" % keras_name)

--- a/onnx2keras/utils.py
+++ b/onnx2keras/utils.py
@@ -1,5 +1,6 @@
 import numpy as np
 from tensorflow import keras
+import torch
 
 
 def is_numpy(obj):
@@ -45,6 +46,37 @@ def ensure_tf_type(obj, fake_input_layer=None, name=None):
         return lambda_layer(fake_input_layer)
     else:
         return obj
+
+
+def count_params_torch(model: torch.nn.Module, trainable_only: bool = False) -> int:
+    """Count the total number of scalars composing the parameters of PyTorch model."""
+
+    if trainable_only:
+        cond = lambda p: p.requires_grad
+    else:
+        cond = lambda p: True
+
+    return sum(p.numel() for p in model.parameters() if cond(p))
+
+
+def count_params_keras(model: keras.Model, trainable_only: bool = False) -> int:
+    # source: https://github.com/keras-team/keras/blob/master/keras/utils/layer_utils.py
+    # Adapted from keras.utils.layer_utils.layer_utils, part of the private API of keras.utils
+    """Count the total number of scalars composing the weights of Keras model."""
+
+    if trainable_only:
+        weights = model.trainable_weights
+    else:
+        weights = model.weights
+
+    unique_weights = {id(w): w for w in weights}.values()
+    # Ignore TrackableWeightHandlers, which will not have a shape defined.
+    unique_weights = [w for w in unique_weights if hasattr(w, 'shape')]
+    weight_shapes = [w.shape.as_list() for w in unique_weights]
+    standardized_weight_shapes = [
+      [0 if w_i is None else w_i for w_i in w] for w in weight_shapes
+    ]
+    return int(sum(np.prod(p) for p in standardized_weight_shapes))
 
 
 def check_torch_keras_error(model, k_model, input_np, epsilon=1e-5, change_ordering=False):

--- a/onnx2keras/utils.py
+++ b/onnx2keras/utils.py
@@ -94,8 +94,9 @@ def check_torch_keras_error(model, k_model, input_np, epsilon=1e-5, change_order
         # change image data format if output shapes are different (e.g. the same for global_avgpool2d)
         _koutput = []
         for i, k in enumerate(keras_output):
+            # @fixme: necessary? see code for keras layer GlobalAvgPool2D to see how is data_format="channels_last" managed
             if k.shape != pytorch_output[i].shape:
-                axes = list(range(len(k.shape)))
+                axes = list(range(len(k.shape)))  # @fixme: axes = k.rank()?
                 axes = axes[0:1] + axes[-1:] + axes[1:-1]
                 k = np.transpose(k, axes)
             _koutput.append(k)
@@ -109,8 +110,12 @@ def check_torch_keras_error(model, k_model, input_np, epsilon=1e-5, change_order
     # reset to previous image_data_format
     keras.backend.set_image_data_format(initial_keras_image_format)
 
+    # assert outputs are all close up to an absolute tolerance
     max_error = 0
     for p, k in zip(pytorch_output, keras_output):
+        # assert shapes are the same
+        assert p.shape == k.shape
+
         error = np.max(np.abs(p - k))
         np.testing.assert_allclose(p, k, atol=epsilon, rtol=0.0)
         if error > max_error:

--- a/test/models/test_resnet18_training_mode.py
+++ b/test/models/test_resnet18_training_mode.py
@@ -1,0 +1,106 @@
+import io
+from typing import Callable
+
+import numpy as np
+import onnx
+import pytest
+import torch
+from torch import nn
+from torchvision.models import resnet18
+
+from onnx2keras import check_torch_keras_error, onnx_to_keras
+from onnx2keras.utils import count_params_keras, count_params_torch
+
+
+def _convert_and_test_model_training_mode(
+    model_class: Callable, include_fc: bool = True
+) -> None:
+
+    # instanciate model and optionally discard fc layer
+    model = model_class()
+    if not include_fc:
+        model.fc = nn.Identity()
+    model.train()
+
+    print(
+        "Created PyTorch model '{}'; param count: {}, trainable param count: {}".format(
+            model_class,
+            count_params_torch(model),
+            count_params_torch(model, trainable_only=True),
+        )
+    )
+
+    # create placeholder input tensors
+    input_np = np.random.randn(1, 3, 224, 224)
+    input_t = torch.from_numpy(input_np).float()
+
+    # export PyTorch model to ONNX
+    input_names = ["input_1"]
+    output_names = ["output_1"]
+
+    temp_f = io.BytesIO()
+    torch.onnx.export(
+        model,
+        input_t,
+        temp_f,
+        verbose=True,
+        export_params=True,
+        input_names=input_names,
+        output_names=output_names,
+        opset_version=14,
+        do_constant_folding=False,
+        dynamic_axes={
+            "input_1": {0: "batch_size"},
+            "output_1": {0: "batch_size"},
+        },
+        training=torch.onnx.TrainingMode.TRAINING,
+    )
+    temp_f.seek(0)
+    onnx_model = onnx.load(temp_f)
+
+    # check the ONNX model
+    onnx.checker.check_model(onnx_model)
+
+    # export ONNX model to TF Keras
+    change_ordering = True
+    k_model = onnx_to_keras(
+        onnx_model,
+        input_names=input_names,
+        input_shapes=[(3, None, None)],
+        name_policy=None,
+        verbose=True,
+        change_ordering=change_ordering,
+    )
+
+    # Assert outputs are all close up to an absolute tolerance or 1e-04
+    # Note: max absolute difference is expected to be around 5e-05 if `include_fc=False`, that's
+    # why we increase `epsilon` from 1e-5 to 1e-4
+    model.eval()
+
+    error = check_torch_keras_error(
+        model,
+        k_model,
+        input_np,
+        epsilon=1e-4,
+        change_ordering=change_ordering,
+    )
+    print(f"Outputs Max Error: {error}")
+
+    # Verify the number of trainable parameters is the same
+    # Note: we don't verify that the number of total parameters is the same
+    # because there are possible differences:
+    # for particular layers such as BatchNorm, some variables are not considered parameters
+    # in PyTorch ('buffers', instead), while they are in Keras.
+
+    n_trainable_params = count_params_torch(model, trainable_only=True)
+    k_n_trainable_params = count_params_keras(k_model, trainable_only=True)
+
+    assert n_trainable_params == k_n_trainable_params
+
+    k_model.summary()
+
+
+@pytest.mark.parametrize("include_fc", [True, False])
+def test_resnet18(include_fc: bool):
+
+    _convert_and_test_model_training_mode(resnet18, include_fc=include_fc)

